api/src/main/java/org/jenkinsci/plugins/workflow/flow/FlowExecution.java

```diff
@@ -158,7 +158,25 @@ public String getUrl() throws IOException {
      */
     public abstract void interrupt(Result r, CauseOfInterruption... causes) throws IOException, InterruptedException;
 
-    public abstract void addListener(GraphListener listener);
+    @Deprecated
+    public void addListener(GraphListener listener) {
+        addListener(listener, false);
+    }
+
+    /**
+     * Add a listener to changes in the flow graph structure.
+     * @param listener a listener to add
+     * @param synchronous true to receive events immediately as they occur (you must be very careful not to acquire locks or block); false to receive notification as soon as possible
+     */
+    public /*abstract*/ void addListener(GraphListener listener, boolean synchronous) {
```

Comment:
I also considered adding a

```java
public interface GraphListener {
    // …
    interface Synchronous extends GraphListener {}
}
```

and wound up going up with a flag on the listener instead, but if you think this is more intuitive as a marker on the listener then it should be easy to change it that way.